File: al-leetcode/uncategorized/dp_matrix-block-sum.js
/**
 * @param {number[][]} mat
 * @param {number} k
 * @return {number[][]}
 */
var matrixBlockSum = function(mat, k) {
    // Input: mat = [[1,2,3],[4,5,6],[7,8,9]], k = 1
    // Output: [[12,21,16],[27,45,33],[24,39,28]]
    // 
    // O(n^2)
    // O(n^2)
    // dp

    // Given a m x n matrix mat and an integer k, return a matrix answer where each answer[i][j] is the sum of all elements mat[r][c] for:
    // i - k <= r <= i + k,
    // j - k <= c <= j + k, and
    // (r, c) is a valid position in the matrix.

    if (mat.length === 0) {
        return [];
    }
    let hashMap = {};
    let result = [];
    // Build hashMap
    for (let i = 0; i < mat.length; i++) {
        for (let j = 0; j < mat[i].length; j++) {
            let key = i + ',' + j;
            hashMap[key] = mat[i][j];
        }
    }
    // Build result
    for (let i = 0; i < mat.length; i++) {
        let row = [];
        for (let j = 0; j < mat[i].length; j++) {
            let sum = 0;
            for (let r = i - k; r <= i + k; r++) {
                for (let c = j - k; c <= j + k; c++) {
                    if (r < 0 || r >= mat.length || c < 0 || c >= mat[i].length) {
                        continue;
                    }
                    let key = r + ',' + c;
                    sum += hashMap[key];
                }
            }
            row.push(sum);
        }
        result.push(row);
    }
    return result;
};

File: al-leetcode/uncategorized/dp_matrix-block-sum.java
class Solution {
    public int[][] matrixBlockSum(int[][] mat, int k) {
        // Input: mat = [[1,2,3],[4,5,6],[7,8,9]], k = 1
        // Output: [[12,21,16],[27,45,33],[24,39,28]]
        int m = mat.length;
        int n = mat[0].length;
        int[][] res = new int[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int left = Math.max(0, j - k);
                int right = Math.min(n - 1, j + k);
                int top = Math.max(0, i - k);
                int bottom = Math.min(m - 1, i + k);
                int sum = 0;
                for (int x = top; x <= bottom; x++) {
                    for (int y = left; y <= right; y++) {
                        sum += mat[x][y];
                    }
                }
                res[i][j] = sum;
            }
        }
        return res;
    }
}